common/unsafe/src/main/java/org/apache/spark/unsafe/Platform.java

```diff
@@ -209,21 +209,25 @@ public static long reallocateMemory(long address, long oldSize, long newSize) {
   }
 
   /**
-   * Uses internal JDK APIs to allocate a DirectByteBuffer while ignoring the JVM's
-   * MaxDirectMemorySize limit (the default limit is too low and we do not want to require users
-   * to increase it).
+   * Allocate a DirectByteBuffer, potentially bypassing the JVM's MaxDirectMemorySize limit.
    */
   public static ByteBuffer allocateDirectBuffer(int size) {
     try {
+      if (CLEANER_CREATE_METHOD == null) {
+        // Can't set a Cleaner (see comments on field), so need to allocate via normal Java APIs
+        return ByteBuffer.allocateDirect(size);
```

Comment:
try /catch OOM and log a message on setting MaxDirectMemorySize?

Reply:
OK, it does look like it will throw OutOfMemoryError here if off-heap allocation fails. That's a decent hint, yes.